### PR TITLE
Add a function which show the statistics infomation during the trace query 

### PR DESCRIPTION
**Pull request by constanine**
The display effect is shown below:
skywalking-trace-statistics-render

### CODE REVIEW

**Comment by constanine** [comment]
[https://github.com/constanine/skywalking-rocketbot-ui/blob/boke/website-doc/8.4.0/new%20issue/skywalking-trace-statc.png](url)

**Comment by wu-sheng** [comment]
Could you resolve the conflicts?

File: trace-statistics-info.md
Comment (wu-sheng) [inline trace-statistics-info.md:7]: You don't have to add this, and also png file. We don't need to increase the package size.

**Comment by wu-sheng** [comment]
Could you share the analytic rules? Are you grouping them through operations name of spans?

File: trace-statistics-info.md
Comment (constanine) [reply trace-statistics-info.md:7]: Ok，should commit a delete changeset?

Comment (wu-sheng) [reply trace-statistics-info.md:7]: Yes, it is fine.

**Comment by constanine** [comment]
> Could you share the analytic rules? Are you grouping them through operations name of spans?

yeah,you can see the code on src/views/components/common/trace-detail-statistics-table.vue on line 78~155, and we grouping them through endpointName

**Comment by constanine** [comment]
> Could you resolve the conflicts?

sorry,should we merge the changes on master branch in mine fork project?
And then, pull requeset in the way of mime fork project on master branch to apache/skywalking master branch？

**Comment by wu-sheng** [comment]
> > Could you resolve the conflicts?
> 
> 
> 
> sorry,should we merge the changes on master branch in mine fork project?
> 
> And then, pull requeset in the way of mime fork project on master branch to apache/skywalking master branch？

It is fine. Once those are following comments ot improvements.

**Comment by wu-sheng** [comment]
@Fine0830 Please take a look. Logically, this feature makes sense to me.

FYI @apache/skywalking-committers

**Comment by constanine** [comment]
> Could you resolve the conflicts?

I have resolve the confilcts now,and waitting for merge

File: src/views/components/common/trace-detail-statistics-table.vue
Comment (wu-sheng) [inline src/views/components/common/trace-detail-statistics-table.vue:156]: Chinese comments are not allowed. @Fine0830 We need to be careful about this, it seems CI doesn't check this. But the main repo CI will check.

File: src/views/components/common/trace-detail-statistics-table.vue
Comment (wu-sheng) [inline src/views/components/common/trace-detail-statistics-table.vue:77]: Same issue, and why TODO?

File: src/views/components/common/trace-detail-statistics-table.vue
Comment (wu-sheng) [inline src/views/components/common/trace-detail-statistics-table.vue:68]: Meaningless codes.

**Comment by wu-sheng** [comment]
Suggestions
1. It seems the analytics result table is unordered. Could you improve that? Such as ordering by avg time or count desc.
2. License check failed, please check CI logs.
3. From the screenshot, I can't see the entry span or local span. Are you analyzing those?
4. Do you use self-duration as the span's time cost?

**Comment by wu-sheng** [comment]
> I have resolve the confilcts now,and waitting for merge

Be patient, merge requires reviewers' confirmation. And now, it is on vacation.

File: src/views/components/common/trace-detail-statistics-table.vue
Comment (constanine) [reply src/views/components/common/trace-detail-statistics-table.vue:77]: sorry,to remind my partner,and he forget to remove this exegesis,I'll remove it and commit

Comment (wu-sheng) [reply src/views/components/common/trace-detail-statistics-table.vue:77]: Please confirm with your colleague, if this PR gets merged, his name and mail wouldn't show up on the ASF. 
It is better you could ask him to confirm here, publicly. Usually, a pull request should belong to one individual.

Comment (wu-sheng) [reply src/views/components/common/trace-detail-statistics-table.vue:77]: I think his GitHub ID is @Visbin

**Comment by constanine** [comment]
> > I have resolve the confilcts now,and waitting for merge
> 
> Be patient, merge requires reviewers' confirmation. And now, it is on vacation.
Ok, I should apologize for not reviewing some changes with my partner,and  it seems to be a good news which is on vacation.
I also need a time to review our changeset,some details is not so good

**Comment by wu-sheng** [comment]
And also, move the `statistics` button to the last(right) one.

File: src/views/components/common/trace-statistics-table/trace-item.vue
Comment (wu-sheng) [inline src/views/components/common/trace-statistics-table/trace-item.vue:48]: Still have commented codes.

File: trace-statistics-info.md
Comment (wu-sheng) [reply trace-statistics-info.md:7]: This is still not removed.

**Comment by wu-sheng** [comment]
Could you private a new screenshot and give some explanation about how you are going to calculate these metrics on the browser?
Such as, do you use span's duration? Or use the duration minus its children's durations? Do you calculate all spans? Do you base on operation name of span(You said about endpoint, but your screenshot says not)?

**Comment by constanine** [comment]
> Could you private a new screenshot and give some explanation about how you are going to calculate these metrics on the browser?

Sorry,I’ll take new screenshot tomorrow.May I explain some reason on old picture
skywalking-trace-statistics-render-explain

**Comment by constanine** [comment]
> Such as, do you use span's duration? Or use the duration minus its children's durations? Do you calculate all spans? Do you base on operation name of span(You said about endpoint, but your screenshot says not)?

1. We use span's duration,by starttime span'endTime minus span'sstartTime
2. We still use function:changetree() at first,then calculate for statistics. So,I think we use the duration minus its children's durations ( all span just have the same parent span )
3. We just calculate all span in the same trace
4. The grouping is based on span's endpoint

**Comment by wu-sheng** [comment]
> We just calculate all span in the same trace
The grouping is based on span's endpoint

Let me be clear, this thing you called endpoint isn't SkyWalking's endpoint. They are span's operation name. Take a look at the GraphQL protocol.

**Comment by wu-sheng** [comment]
@Fine0830 Please do code review.

File: src/views/components/common/trace-statistics-table/trace-item.vue
Comment (Fine0830) [inline src/views/components/common/trace-statistics-table/trace-item.vue:44]: As above. Please reuse existing components.

**Review by Fine0830 — CHANGES_REQUESTED** [review]

File: src/views/components/common/trace-statistics-table/trace-item.vue
Comment (constanine) [reply src/views/components/common/trace-statistics-table/trace-item.vue]: > As above. Please reuse existing components.

We have merge code in trace.scss,trace-constant.ts.
But in trace-container.vue, it shows different table columns, and use v-if is not a good idead,we decide to rname and see new commit

**Comment by Visbin** [comment]
The latest commits have been added with sorting capabilities.

Support for ascending and descending order.

Includes: MaxTime, MinTime, AvgTime, SumTime, Hits

Hits is the renaming of Count.

image

**Comment by wu-sheng** [comment]
@Fine0830 The screenshot is good to me. Please recheck.

File: src/views/components/trace/trace-detail.vue
Comment (Fine0830) [inline src/views/components/trace/trace-detail.vue:45]: Please use `rk-icon` component instead this

File: src/views/components/common/trace-chart-table/trace-statistics-container.vue
Comment (Fine0830) [inline src/views/components/common/trace-chart-table/trace-statistics-container.vue:32]: Please use `ts`

File: src/views/components/common/trace-chart-table/trace-statistics-item.vue
Comment (Fine0830) [inline src/views/components/common/trace-chart-table/trace-statistics-item.vue:43]: Please use `ts`

File: src/views/components/common/trace-detail-statistics-table.vue
Comment (Fine0830) [reply src/views/components/common/trace-detail-statistics-table.vue:38]: This component is same as common/trace-detail-chart-table . Please reuse existing components rather than creating a same components.

**Review by Fine0830 — CHANGES_REQUESTED** [review]

**Comment by constanine** [comment]
@Fine0830 I'think src/views/components/common/trace-detail-statistics-table.vue should be independent.
First,It does not the same as  common/trace-detail-chart-table, now you can't say or table view is the same. The comopnet to show is different and the aim for calculate is different thing.
Second, we see src/views/components/trace/trace-detail-chart-list.vue and src/views/components/trace/trace-detail-chart-tree.vue there are similar too.

File: src/views/components/common/trace-detail-statistics-table.vue
Comment (Fine0830) [inline src/views/components/common/trace-detail-statistics-table.vue:69]: Could you add a `if` to `return`, when the component isn't the statistics table. And move the function to `trace-detail-chart-table.vue`. Other functions seem to look the same. WDYT @wu-sheng @Jtrust  

Should the statistics table be same as the other three types to support clicking to view trace span logs? @wu-sheng

Comment (wu-sheng) [reply src/views/components/common/trace-detail-statistics-table.vue:69]: @Fine0830 What do you want me to confirm?

Comment (Fine0830) [reply src/views/components/common/trace-detail-statistics-table.vue:69]: uh....I mean, what do you think of this suggestion? And is this function supported?

Comment (wu-sheng) [reply src/views/components/common/trace-detail-statistics-table.vue:69]: Why do you move the statistic page into the span view? This PR is making a statistics for all spans, right? It tries to find the slowest or hot spans, from my understanding.
I am not 100% sure.

Comment (Visbin) [reply src/views/components/common/trace-detail-statistics-table.vue:69]: Yeap. It doesn't seem necessary

Comment (wu-sheng) [reply src/views/components/common/trace-detail-statistics-table.vue:69]: This PR merges the span according to operation name and get statistics data calculated in the browser. So, you need to recheck the algorithm. @Fine0830

**Comment by wu-sheng** [comment]
@Visbin @constanine I still can see both of these accounts submitting codes. Could you share what is the status?

**Comment by constanine** [comment]
> @Visbin @constanine I still can see both of these accounts submitting codes. Could you share what is the status?

I'think the primary goal which does successfully merge the PR . 🤣

**Comment by wu-sheng** [comment]
> > @Visbin @constanine I still can see both of these accounts submitting codes. Could you share what is the status?
> 
> I'think the primary goal which does successfully merge the PR . 🤣

I am asking, are you two individuals?

**Comment by Fine0830** [comment]
There are two problems.
1. JS errors


2. Data calculation errors. From the example, the answer is `Max Time=8040，min time=1, avg=(8040+1) / 2, sum = 8041, hit = 2`.
WechatIMG28
WechatIMG21

**Comment by constanine** [comment]
> > > @Visbin @constanine I still can see both of these accounts submitting codes. Could you share what is the status?
> > 
> > 
> > I'think the primary goal which does successfully merge the PR . 🤣
> 
> I am asking, are you two individuals?

Yes，we are two different individuals,I know what do you mean.
Let’s take another topic,I know to be member of ASF is an honor,but every crown has it’s weight,So,what
 will come if be a member of ASF？
And this is my first PR to apache‘project，also to my collage，so the thing of decide to which one to be ASF should be carefully
To completed is more important，and welcome to give more suggest
By the way，I‘m not good at english，just have passed ect4.So, even this, I abide by the agreement
,which do reply,commit code log all in english.I want to the PR is done,Thanks

**Comment by Visbin** [comment]
> There are two problems.
> 
> 1. JS errors
>    
> 2. Data calculation errors. From the example, the answer is `Max Time=8040，min time=1, avg=(8040+1) / 2, sum = 8041, hit = 2`.
>    WechatIMG28
>    WechatIMG21

Thank you very much for your question. It has been repaired. Please check it again

**Review by Fine0830 — COMMENTED** [review]
1. JS Errors

2.  Please use `ts`. `tslint` can't constrain `js` code.  There are some lint errors in programs.

**Comment by constanine** [comment]
> 1. JS Errors
>    
> 2. Please use `ts`. `tslint` can't constrain `js` code.  There are some lint errors in programs.

Can you elaborate on first question which I see has a error in src/components/rk-page.vue?
What button is it to click?
I see git log there is nothing change in src/components/rk-page.vue,and src/views/components/trace/trace-detail.vue has releated just add a import for components

**Comment by Fine0830** [comment]
> > 1. JS Errors
> >    
> > 2. Please use `ts`. `tslint` can't constrain `js` code.  There are some lint errors in programs.
> 
> Can you elaborate on first question which I see has a error in src/components/rk-page.vue?
> What button is it to click?
> I see git log there is nothing change in src/components/rk-page.vue,and src/views/components/trace/trace-detail.vue has releated just add a import for components

When I turn the page, the error was reproduced

**Comment by constanine** [comment]
> 2. Please use `ts`. `tslint` can't constrain `js` code.  There are some lint errors in programs.
Here my code check result show no tslint error...
image

**Comment by constanine** [comment]
> > > 1. JS Errors
> > >    
> > > 2. Please use `ts`. `tslint` can't constrain `js` code.  There are some lint errors in programs.
> > 
> > 
> > Can you elaborate on first question which I see has a error in src/components/rk-page.vue?
> > What button is it to click?
> > I see git log there is nothing change in src/components/rk-page.vue,and src/views/components/trace/trace-detail.vue has releated just add a import for components
> 
> When I turn the page, the error was reproduced
> 

The picture which you show I haven't found our change such as there is no button of statistics, and I have tried turn the page，there is no error

https://user-images.githubusercontent.com/7012545/118632840-6727b400-b803-11eb-8a04-4ca8895463d7.mp4
image

File: src/views/components/common/trace-detail-statistics-table.vue
Comment (Fine0830) [inline src/views/components/common/trace-detail-statistics-table.vue:71]: Identifier 'map' is never reassigned; use 'const' instead of 'let'.

File: src/views/components/common/trace-detail-statistics-table.vue
Comment (Fine0830) [inline src/views/components/common/trace-detail-statistics-table.vue:80]: Identifier 'arr' is never reassigned; use 'const' instead of 'let'.

**Review by Fine0830 — CHANGES_REQUESTED** [review]
I only write several comments here. You can't find lint errors as you use `js` to code, rather than `ts`.  **Please use `ts`.**
`<script lang="ts">.....</script>`

I always find this error.

**Comment by constanine** [comment]
> I only write several comments here. You can't find lint errors as you use `js` to code, rather than `ts`. **Please use `ts`.**
> `<script lang="ts">.....</script>`
> 
> When I input page number, here always find this error.
> 
> 
> demo env works well
> 

Are you sure to be `<script lang="ts">.....</script>`? I see in the master branch,trace-detail-chart-tree.vue、trace-detail-chart-list.vue and trace-detail-chart-table.vue all use `<script lang="js">.....</script>`,I don't want to break some rule, as other do... 
Or you‘re ready to change these vue file to use `<script lang="ts">.....</script>`

**Comment by constanine** [comment]
> I only write several comments here. You can't find lint errors as you use `js` to code, rather than `ts`. **Please use `ts`.**
> `<script lang="ts">.....</script>`
> 
> When I input page number, here always find this error.
> 
> 
> demo env works well
> 

 `<script lang="ts">.....</script>`

**Comment by constanine** [comment]
> I only write several comments here. You can't find lint errors as you use `js` to code, rather than `ts`. **Please use `ts`.**
> `<script lang="ts">.....</script>`
> 
> When I input page number, here always find this error.
> 
> 
> demo env works well
> 

I see src/views/components/common/trace-detail-chart-table.vue,src/views/components/trace/trace-detail-chart-list.vue,
src/views/components/trace/trace-detail-chart-tree.vue all just use `<script lang="js">.....</script>` not  `<script lang="ts">.....</script>`, Are you sure to use `<script lang="ts">.....</script>`? And are you ready to have a plan which change them to use  `<script lang="ts">.....</script>`?

**Comment by wu-sheng** [comment]
I don't know UI details. @Fine0830 makes the call.

@constanine I need to remind you, from reviewer's perspective, to focus on the codes you proposed, don't challenge the existing codes not changed. Those are 2 separate things.

**Comment by constanine** [comment]
> I only write several comments here. You can't find lint errors as you use `js` to code, rather than `ts`. **Please use `ts`.**
> `<script lang="ts">.....</script>`
> 
> When I input page number, here always find this error.
> 
> 
> demo env works well
> 

I see src/views/components/common/trace-detail-chart-table.vue,src/views/components/trace/trace-detail-chart-list.vue,
src/views/components/trace/trace-detail-chart-tree.vue and lmost all the *.vue just use `<script lang="js">.....</script>` not  `<script lang="ts">.....</script>`, Are you sure to use `<script lang="ts">.....</script>`? And are you ready to have a plan which change them to use  `<script lang="ts">.....</script>`? 
And if I use  `<script lang="ts">.....</script>` it need all the *.vue to use  `<script lang="ts">.....</script>`! I think this should be an other PR, not mime ! 
How do you think @

> I don't know UI details. @Fine0830 makes the call.
> 
> @constanine I need to remind you, from reviewer's perspective, to focus on the codes you proposed, don't challenge the existing codes not changed. Those are 2 separate things.

Beasuse of almost views/component/*.vue need to change, if is ts we can't debugger in browser,and if type is js,we can debugger in browser, and not I think this ui juse ui for browser not for server, this project just build a  dist and player with java web application ,so it need ts? 
image

**Comment by Fine0830** [comment]
There are most components that use `ts` to code in the project, except trace components, etc. We expect the new components to be implemented with `ts`. The project use `tslint` to constrain code, so code with `ts` is better. Some components that use `js` are legacy issues, they will be refactored.

**Comment by wu-sheng** [comment]
> Beasuse of almost views/component/*.vue need to change, if is ts we can't debugger in browser,and if type is js,we can debugger in browser, and not I think this ui juse ui for browser not for server, this project just build a dist and player with java web application ,so it need ts?

As I said, I am not a FE. I am not the right one to talk to. But I have said, follow the reviewer's comments. Rather than referring to other things.

**Comment by constanine** [comment]
@Fine0830  as you wash ,I change the code ,use ts, please review later

File: src/views/components/common/trace-chart-table/trace-container.vue
Comment (Fine0830) [inline src/views/components/common/trace-chart-table/trace-container.vue:78]:
```suggestion
      if (this.type === 'statistics') {
        return;
      }
      const drag: any = this.$refs.dragger;
      ......
```

**Review by Fine0830 — CHANGES_REQUESTED** [review]
JS Errors

File: src/views/components/common/trace-detail-statistics-table.vue
Comment (constanine) [reply src/views/components/common/trace-detail-statistics-table.vue:42]: > Please delete this.



/* eslint-disable */  /* tslint:disable */
 Is this？ Not import script?

**Comment by constanine** [comment]
> JS Errors
> 
> 
This error I can't Reproduce,And I review the change log ,I never change trace-table.vue...

**Comment by constanine** [comment]
> JS Errors
> 
> 
cloud you give the data which cause this error ? My data is running Ok with no error

**Comment by Fine0830** [comment]
> > JS Errors
> > 
> > This error I can't Reproduce,And I review the change log ,I never change trace-table.vue...
> > Clicking on a button, or turnning on a page, can't repeat a error report.
> > Except this, other buges has been fixed, please review later,thanks

first error:
step one: visit http://demo.skywalking.apache.org/trace
step two: click the button of statistics

second error:
when you input page number, the error will be reproduced.

**Comment by constanine** [comment]
> > > JS Errors
> > > 
> > > This error I can't Reproduce,And I review the change log ,I never change trace-table.vue...
> > > Clicking on a button, or turnning on a page, can't repeat a error report.
> > > Except this, other buges has been fixed, please review later,thanks
> 
> first error:
> step one: visit http://demo.skywalking.apache.org/trace
> step two: click the button of statistics
> 
> second error:
> when you input page number, the error will be reproduced.

visit http://demo.skywalking.apache.org/trace need a account... Can you give one？

**Comment by constanine** [comment]
> > > JS Errors
> > > 
> > > This error I can't Reproduce,And I review the change log ,I never change trace-table.vue...
> > > Clicking on a button, or turnning on a page, can't repeat a error report.
> > > Except this, other buges has been fixed, please review later,thanks
> 
> first error:
> step one: visit http://demo.skywalking.apache.org/trace
> step two: click the button of statistics
> 
> second error:
> when you input page number, the error will be reproduced.

there 2 erros are fixed，please review later ,thanks

**Review by Fine0830 — COMMENTED** [review]
Codes are fine. I have some doubts about the result of `statistics`. After span data merging, there should be only two, is right? Please check @wu-sheng

**Comment by wu-sheng** [comment]
> After span data merging, there should be only two, is right?

@Fine0830 What do you mean 2? All operation names should be visualized, only the same names should be merged.

**Comment by Fine0830** [comment]
> > After span data merging, there should be only two, is right?
> 
> @Fine0830 What do you mean 2? All operation names should be visualized, only the same names should be merged.

Alright. Before, I thought there should be two datas, ` /projectA/test` and `test.skywalking.springcloud.test.projectd.MessageConustumer`. If so, I have no questions.

**Review by Fine0830 — APPROVED** [review]
LGTM

**Comment by wu-sheng** [comment]
@constanine You need to update the submodule of the main repo, to make this feature available in the release.

**Pull request merged**
